src/couch_stats/test/eunit/csrt_logger_tests.erl

```diff
@@ -0,0 +1,368 @@
+                %% and changes_returned but the latter must be <= the former
+                ('_do_changes', V, Acc) ->
+                    case V of
+                        true ->
+                            Processed = R(),
+                            Returned = (one_of([0, 0, 1, Processed, rand:uniform(Processed)]))(),
+                            maps:put(
+                                changes_processed,
+                                Processed,
+                                maps:put(changes_returned, Returned, Acc)
+                            );
+                        _ ->
+                            Acc
+                    end;
+                (K, F, Acc) when is_function(F) ->
+                    maps:put(K, F(), Acc);
+                (K, V, Acc) ->
+                    maps:put(K, V, Acc)
+            end,
+            #{},
+            Opts
+        )
+    ).
+
+rctxs() ->
+    [rctx_gen() || _ <- lists:seq(1, ?RCTX_COUNT)].
+
+t_do_report(#{rctx := Rctx}) ->
+    JRctx = csrt_util:to_json(Rctx),
+    ReportName = "foo",
+    ok = meck:new(couch_log),
+    ok = meck:expect(couch_log, report, fun(_, _) -> true end),
+    ?assert(csrt_logger:do_report(ReportName, Rctx), "CSRT _logger:do_report " ++ ReportName),
+    ?assert(meck:validate(couch_log), "CSRT do_report"),
+    ?assert(meck:validate(couch_log), "CSRT validate couch_log"),
+    ?assert(
+        meck:called(couch_log, report, [ReportName, JRctx]),
+        "CSRT couch_log:report"
+    ),
+    ok = meck:unload(couch_log).
+
+t_do_lifetime_report(#{rctx := Rctx}) ->
+    JRctx = csrt_util:to_json(Rctx),
+    ReportName = "csrt-pid-usage-lifetime",
+    ok = meck:new(couch_log),
+    ok = meck:expect(couch_log, report, fun(_, _) -> true end),
+    ?assert(
+        csrt_logger:do_lifetime_report(Rctx),
+        "CSRT _logger:do_report " ++ ReportName
+    ),
+    ?assert(meck:validate(couch_log), "CSRT validate couch_log"),
+    ?assert(
+        meck:called(couch_log, report, [ReportName, JRctx]),
+        "CSRT couch_log:report"
+    ),
+    ok = meck:unload(couch_log).
+
+t_do_status_report(#{rctx := Rctx}) ->
+    JRctx = csrt_util:to_json(Rctx),
+    ReportName = "csrt-pid-usage-status",
+    ok = meck:new(couch_log),
+    ok = meck:expect(couch_log, report, fun(_, _) -> true end),
+    ?assert(csrt_logger:do_status_report(Rctx), "csrt_logger:do_ " ++ ReportName),
+    ?assert(meck:validate(couch_log), "CSRT validate couch_log"),
+    ?assert(
+        meck:called(couch_log, report, [ReportName, JRctx]),
+        "CSRT couch_log:report"
+    ),
+    ok = meck:unload(couch_log).
+
+t_matcher_on_dbname(#{rctx := _Rctx, rctxs := Rctxs0}) ->
+    %% Make sure we have at least one match
+    Rctxs = [rctx_gen(#{dbname => <<"foo">>}) | Rctxs0],
+    ?assertEqual(
+        lists:sort(lists:filter(matcher_on(dbname, <<"foo">>), Rctxs)),
+        lists:sort(lists:filter(matcher_for_csrt("dbname"), Rctxs)),
+        "Dbname matcher on <<\"foo\">>"
```

Comment:
Thank you for using `?assertEqual/3` form of an assert functions. We should use them more. This gives us opportunity to provide clean error messages on failures.